src/nxt_cgroup.c

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright (C) Andrew Clayton
+ * Copyright (C) F5, Inc.
+ */
+
+#include <nxt_main.h>
+#include <nxt_process.h>
+#include <nxt_fs.h>
+#include <nxt_cgroup.h>
+
+#include <stdio.h>
+#include <string.h>
+#include <limits.h>
+
+
+nxt_int_t
+nxt_cgroup_proc_add(const nxt_task_t *task, nxt_process_t *process)
+{
+    FILE  *fp;
+    char  cgprocs[PATH_MAX];
+    char  cgpath[PATH_MAX - 13];    /* - 13 for '/cgroup.procs' */
+
+    if (task->thread->runtime->type != NXT_PROCESS_MAIN
+        || nxt_process_type(process) != NXT_PROCESS_PROTOTYPE
+        || process->isolation.cgroup.name == NULL)
+    {
+        return NXT_OK;
+    }
+
+    snprintf(cgpath, sizeof(cgpath), NXT_CGROUP_ROOT "/%s",
```

Comment:
As the `isolation.cgroup.name` config can contain slashes, this can lead to the creation of a `cgroup.procs` into arbitrary directories, no? Eg.:
```
"isolation": {
     "cgroup": {
         "name": "/../../../../root/hello-there"
     }
}
```
If that's the case, I think it's reasonable to call `nxt_realpath()` and check if the returned value still has the `NXT_CGROUP_ROOT` prefix and fail otherwise.

Reply:
We could check that in the conf validation.

Reply:
We can't really use realpath(3) here as there's a good chance the path we're checking doesn't exist and realpath will throw an error.

Now the Linux [man page](https://man7.org/linux/man-pages/man3/realpath.3.html) says under RETURN VALUE

> ...
> Otherwise, it returns NULL, the contents of the array resolved_path are
> undefined, and errno is set to indicate the error.

Although it does say further down under GNU extensions

>If the call fails with either EACCES or ENOENT and resolved_path is not
>NULL, then the prefix of path that is not readable or does not exist is
>returned in resolved_path

Which would work for us, but obviously we can't rely on that behaviour.

However a couple of strstr(3)'s checking for "/./" or "/../" in the path seems to do the trick and should be fine in this use case. Technically we don't need to check for "/./" but realpath(3) gets rid of it and I see no reason why it would want to be used here...

Reply:
You still can use `realpath(3)` but you need to remove path components until it reaches an existent directory, then you check if this directory is under `NXT_CGROUP_ROOT`:
```
realpath("/sys/fs/cgroup/../../../../../root/non-existent-dir") -> fail
realpath("/sys/fs/cgroup/../../../../../root") -> /root
```

another example:
```
realpath("/sys/fs/cgroup/../test") -> fail
realpath("/sys/fs/cgroup/..") -> /sys/fs
```

Filters can be bypassed but if you go in this direction then maybe also fix the case for `cgroup.name=..` as it will be valid and lead to directory `/sys/fs`. A `strstr(path, "/..")` is an option but invalidate paths like `..something` ...

I personally like the `realpath(3)` approach because the `cgroup.name` allows a path and `/../` is a valid path component. In the end what you want to enforce is: _whatever path the user chooses, it should be inside `NXT_CGROUP_ROOT`_.

Reply:
Hmm, looping realpath(3) removing components until we don't fail could work, just not sure it's worth the extra complexity in this case.

> valid and lead to directory /sys/fs. A strstr(path, "/..") is an option but invalidate paths like ..something

'/sys/fs' would be denied with a
```C
realpath(path, rpath); /* rpath = '/sys/fs' */
if (strncmp(rpath, NXT_CGROUP_ROOT, strlen(NXT_CGROUP_ROOT)) != 0)
```

With strstr(path, "/../") (note the trailing '/' and path is the _full_ path including NXT_CGROUP_ROOT), you can quite happily specify things like "..fundir../app1"
 
> I personally like the realpath(3) approach because the cgroup.name allows a path and /../ is a valid path

Think of it more as a 'name' (it is cgroup.name and not cgroup.path after all...) that happens to match to an underlying path. I see no reason for the name to have '/../' in it.

Reply:
> We can't really use realpath(3) here as there's a good chance the path we're checking doesn't exist and realpath will throw an error.
> 
> Now the Linux [man page](https://man7.org/linux/man-pages/man3/realpath.3.html) says under RETURN VALUE
> 
> > ...
> > Otherwise, it returns NULL, the contents of the array resolved_path are
> > undefined, and errno is set to indicate the error.
> 
> Although it does say further down under GNU extensions
> 
> > If the call fails with either EACCES or ENOENT and resolved_path is not
> > NULL, then the prefix of path that is not readable or does not exist is
> > returned in resolved_path
> 
> Which would work for us, but obviously we can't rely on that behaviour.

We would be looking at *canonicalize_file_name(3)*.  Sadly it's a GNU extension.  Not present in any other libc.  Not even musl.

> However a couple of strstr(3)'s checking for "/./" or "/../" in the path seems to do the trick and should be fine in this use case. Technically we don't need to check for "/./" but realpath(3) gets rid of it and I see no reason why it would want to be used here...

That's not going to be the easiest, because what if the path starts with `../` (no leading slash), or ends with `/..` (no trailing slash)?  Maybe we need to implement *canonicalize_file_name(3)* internally?

Reply:
But if you manage to make it work with *strstr(3)*, it's simpler, so I'd be fine with it.

Reply:
BTW, I just asked in the musl mailing list if they would consider adding canonicalize_file_name(3).

Reply:
Ahhh, you managed it by adding a leading and trailing slash yourself.  That's perfect.

Reply:
@alejandro-colomar 

Yeah, I saw canonicalize_file_name(3), but as you say, it's a GNU extension.

But then again cgroups only work on Linux anyway, but again as you say, there is more than just glibc...

Yeah, before checking the path I add cgroup.name to NXT_CGROUP_ROOT to make the full path so it will catch things like `cgroup.name = "../escape"`.

But good point about a trailing '/..', I don't yet add add a trailing '/' to the path for the doing the check, but will do now!

Reply:
The `canonicalize_file_name(3)` has the same behavior of `realpath(3)`, it actually just a convenience function AFAIK.
`canonicalize_file_name(path)` is the same as `realpath(path, NULL)`.

From the security point of view, I think the `realpath(3)` approach is safer but if you also fix the `name=..` then maybe this is good enough :+1:

Reply:
Oh, you're right.  I misunderstood what *canonicalize_file_name(3)* does.  It also fails for nonexistent paths:

```c
alx@asus5775:~/tmp$ cat cfn.c 
#define _GNU_SOURCE
#include <stdio.h>
#include <stdlib.h>

int main(void)
{
	printf("<%s>\n", canonicalize_file_name("/etc/../foo/bar/./baz"));
}
```
```sh
alx@asus5775:~/tmp$ cc -Wall -Wextra cfn.c 
alx@asus5775:~/tmp$ ./a.out 
<(null)>
```

So we either loop over *realpath(3)*, or make it simple with *strstr(3)*.  I prefer simplicity.  If we completely prohibit `/../`, it should be impossible to trick us, I think.

Reply:
Just for the sake of completeness, I wanted to make a function that loops over realpath(1) correctly.  Coincidently, a bug has been reported to GNU make(1) in the last days, which also relates to this functionality.

So I decided to write it.  Here's the function, and a program resembling *realpath(1)* that uses it.  It doesn't even yet get rid of `/../` after the begining of the first non-existent pathname segment, and you can already see the complexity.  I think we better avoid it.

```c
$ cat normalizepath.c 
#include <bsd/string.h>
#include <err.h>
#include <stddef.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>


char *canonicalizepath(const char *path);


int
main(int argc, char *argv[])
{
	int   ret;
	char *p;

	if (argc < 2)
		errx(EXIT_FAILURE, "missing operand");

	ret = EXIT_SUCCESS;

	for (ptrdiff_t i = 1; i < argc; i++) {
		p = canonicalizepath(argv[i]);
		if (!p) {
			ret = EXIT_FAILURE;
			warn("%s", argv[1]);
		} else {
			puts(p);
			free(p);
		}
	}

	exit(ret);
}


char *
canonicalizepath(const char *path)
{
	char *cpy, *end, *sep, *p, *ret;
	size_t len;

	cpy = strdup(path);
	if (!cpy)
		return NULL;

	end = cpy + strlen(cpy);
	sep = end;

	do {
		*sep = '\0';

		p = realpath(cpy, NULL);
		if (p)
			goto found;

		sep = strrchr(cpy, '/');
	} while (sep);

	sep = cpy;
found:
	len = end - sep + 1;
	if (p)
		len += strlen(p);

	ret = malloc(len);
	if (!ret)
		goto fail;

	ret[0] = '\0';

	if (p)
		strlcat(ret, p, len);
	for (char *q = sep; q < end; q++) {
		if (*q == '\0')
			*q = '/';
	}
	strlcat(ret, sep, len);

fail:
	free(cpy);
	return ret;
}
```

```sh
alx@asus5775:~/tmp$ ./normalizepath .
/home/alx/tmp
alx@asus5775:~/tmp$ ./normalizepath ./foo/../asd
/home/alx/tmp/foo/../asd
alx@asus5775:~/tmp$ ./normalizepath ./asd
/home/alx/tmp/asd
alx@asus5775:~/tmp$ ./normalizepath asd
asd
alx@asus5775:~/tmp$ ./normalizepath normalizepath
/home/alx/tmp/normalizepath
alx@asus5775:~/tmp$ ./normalizepath /home/../tmp/../../etc/
/etc
```

Reply:
Compile the above with:

```sh
cc -Wall -Wextra normalizepath.c $(pkg-config --cflags --libs libbsd) -o normalizepath
```

Reply:
Thanks. Yeah, agreed, for this use case simply checking for '/../' is sufficient and I see no reason why you'd want/need to specify that in a path here.

Reply:
@alejandro-colomar @ac000 simplicity is a good point :+1: so yeah the `/../` seems good enough.